File: src/tokens.py
"""
Token definitions for the BASIC lexer.
"""
from enum import Enum, auto
from dataclasses import dataclass
from typing import Any, Optional


class TokenType(Enum):
    # Literals and identifiers
    NUMBER = auto()
    STRING = auto()
    IDENT = auto()
    # Line number (leading number at start of line)
    LINENUM = auto()
    # Keywords
    PRINT = auto()
    LET = auto()
    INPUT = auto()
    IF = auto()
    THEN = auto()
    ELSE = auto()
    END = auto()
    GOTO = auto()
    GOSUB = auto()
    RETURN = auto()
    FOR = auto()
    TO = auto()
    STEP = auto()
    NEXT = auto()
    REM = auto()
    # Operators
    PLUS = auto()
    MINUS = auto()
    STAR = auto()
    SLASH = auto()
    EQ = auto()       # = (assignment or equality)
    LT = auto()
    LE = auto()
    GT = auto()
    GE = auto()
    NE = auto()       # <>
    LPAREN = auto()
    RPAREN = auto()
    COMMA = auto()
    COLON = auto()
    # Special
    NEWLINE = auto()
    EOF = auto()


KEYWORDS = {
    "PRINT": TokenType.PRINT,
    "LET": TokenType.LET,
    "INPUT": TokenType.INPUT,
    "IF": TokenType.IF,
    "THEN": TokenType.THEN,
    "ELSE": TokenType.ELSE,
    "END": TokenType.END,
    "GOTO": TokenType.GOTO,
    "GOSUB": TokenType.GOSUB,
    "RETURN": TokenType.RETURN,
    "FOR": TokenType.FOR,
    "TO": TokenType.TO,
    "STEP": TokenType.STEP,
    "NEXT": TokenType.NEXT,
    "REM": TokenType.REM,
}


@dataclass
class Token:
    type: TokenType
    value: Any
    line: int
    column: int

    def __repr__(self) -> str:
        return f"Token({self.type.name}, {self.value!r}, {self.line}:{self.column})"

File: webapp/src/compiler/tokens.ts
export const TokenType = {
  NUMBER: 0, STRING: 1, IDENT: 2, LINENUM: 3,
  PRINT: 4, LET: 5, INPUT: 6, IF: 7, THEN: 8, ELSE: 9, END: 10, GOTO: 11, GOSUB: 12, RETURN: 13,
  FOR: 14, TO: 15, STEP: 16, NEXT: 17, REM: 18,
  PLUS: 19, MINUS: 20, STAR: 21, SLASH: 22, EQ: 23, LT: 24, LE: 25, GT: 26, GE: 27, NE: 28,
  LPAREN: 29, RPAREN: 30, COMMA: 31, COLON: 32, NEWLINE: 33, EOF: 34,
} as const;

export type TokenTypeId = (typeof TokenType)[keyof typeof TokenType];

export interface Token {
  type: TokenTypeId;
  value: string | number | null;
  line: number;
  column: number;
}

export const KEYWORDS: Record<string, TokenTypeId> = {
  PRINT: TokenType.PRINT, LET: TokenType.LET, INPUT: TokenType.INPUT,
  IF: TokenType.IF, THEN: TokenType.THEN, ELSE: TokenType.ELSE, END: TokenType.END,
  GOTO: TokenType.GOTO, GOSUB: TokenType.GOSUB, RETURN: TokenType.RETURN,
  FOR: TokenType.FOR, TO: TokenType.TO, STEP: TokenType.STEP, NEXT: TokenType.NEXT,
  REM: TokenType.REM,
};
